### PR TITLE
New eligibility version

### PR DESCRIPTION
**Pull request by JulianRoesner**
# Short Description
This adds the new logic to the eligibility data model. However, this does not touch the steps or the flow in any way. The new way to test the validity of the eligibility is to check for the corresponding data model on a step and use this to get feedback on the data.

If you need to have a look at the logic, you can find it [here](https://www.figma.com/file/8wdB4SGxroQ4hBrO8lbjJ4/Steuerlotse-UI?node-id=2444%3A3091).

# Changes
- Add a new data type RecursiveDataModel: This allows to test recursively the nested data models
- Add all the new logic to the eligibility check by introducing data models for all the steps
- Add tests for both

# Feedback
- Do you have any idea how we might get rid of the `one_previous_field_has_to_be_set` in all the children of the RecursiveDataModel?
- Do you think the tests make sense?
- Do you see anything that I've missed?
- Do you think it makes sense to keep developing on this branch and merge it when everything is done?

### CODE REVIEW

**Review by frederike-ramin — COMMENTED** [review]
Wow. OMG. I *love* the tests!! 😍
I'm excited to see how we use those models in the steps. :)

> Do you have any idea how we might get rid of the one_previous_field_has_to_be_set in all the children of the RecursiveDataModel?

Can we maybe have a class variable `last_previous_field` in all subclasses and just call the validator with that name?

> Do you think it makes sense to keep developing on this branch and merge it when everything is done?

I slightly prefer to merge this first to maintain readibility, but whatever suits you best.

File: webapp/tests/model/test_eligibility_data.py [hunk 1]
```diff
@@ -0,0 +1,1688 @@
...
+        except ValidationError:
+            self.fail("FamilienStandEligibilityData.parse_obj should not raise validation error")
+
+
+class TestLongSeparateLivingEligibilityData(unittest.TestCase):
+
+    def test_if_marital_status_valid_and_separated_since_last_year_no_then_raise_validation_error(self):
+        non_valid_data = {'separated_since_last_year': 'no'}
+        with patch('app.model.eligibility_data.MarriedEligibilityData.parse_obj'):
+            self.assertRaises(ValidationError, LongSeparateLivingEligibilityData.parse_obj, non_valid_data)
```
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/tests/model/test_eligibility_data.py:87]: We talked about it and I got the logic behind the DecisionSteps. 👍

File: webapp/app/model/recursive_data.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    def _update_input(self, data: Any) -> Any:
+        enriched_data = copy.deepcopy(data)
+
+        fields = list(self.__class__.__fields__.values())
+        self.__class__._previous_fields = []
+
+        for field in fields:
+            if issubclass(field.type_, BaseModel):
+                self.__class__._previous_fields.append(field.name)
+                self._set_data_for_previous_field(enriched_data, field.name, field.type_)
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/model/recursive_data.py:30]: By checking the class of the field. Of course, this would fail if the actual field of the class would be of the type BaseModel

File: webapp/app/model/eligibility_data.py [hunk 3]
```diff
@@ -60,3 +80,351 @@ def declarations_must_be_set_no(cls, v, field):
...
+class DivorcedEligibilityData(BaseModel):
+    marital_status: str
+
+    @validator('marital_status')
+    def must_be_divorced(cls, v):
+
+        if v not in 'divorced':
+            raise ValueError
+
+        return v
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/model/eligibility_data.py:142]: I'll change it :)

File: webapp/tests/model/test_eligibility_data.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/tests/model/test_eligibility_data.py:87]: I'll add more documentation for that. The reason is that we use the data types also to differentiate which branch we are taking (if the data is valid stay on the main branch and if it is not valid take the alternative branch). Unfortunately, this is not the most intuitive concept. However, it is quite useful because it makes using the data types for the steps easier.

I'll also rename the `LongSeparateLivingEligibilityData` to `SeparatedEligibilityData` and `ShortSeparateLivingEligibilityData` to `NotSeparatedEligibilityData`

File: webapp/tests/model/test_eligibility_data.py [hunk 4]
```diff
@@ -0,0 +1,1688 @@
...
+
+    def test_if_marital_status_invalid_and_separated_since_last_year_yes_then_raise_validation_error(self):
+        valid_data = {'separated_since_last_year': 'yes'}
+        with patch('app.model.eligibility_data.MarriedEligibilityData.parse_obj',
+                   MagicMock(side_effect=ValidationError([], MarriedEligibilityData))):
+            self.assertRaises(ValidationError, LongSeparateLivingEligibilityData.parse_obj, valid_data)
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/tests/model/test_eligibility_data.py:93]: Yes, renaming is a good idea. However, at this point we do not care if `marital_status` is married, we only care that `MarriedEligibilityData` is valid. Therefore, I would propose: `test_if_married_data_is_invalid_and_separated_yes_then_raise_validation_error` and so on.

I think that we should keep the data model clean as this is our way to check if something is wrong with the data. Therefore, I would say that we do care about the status of the separated question at this point.

File: webapp/tests/model/test_recursive_data.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+    def test_if_two_previous_fields_and_both_raise_value_error_then_raise_validation_error(self):
+        input_data = {'can_jump': True, 'can_fly': True, 'pet_name': 'Yoshi'}
+        CapeMario.should_raise_error = True
+        BaseMario.should_raise_error = True
+        self.assertRaises(ValidationError, self.TwoPreviousMockRecursiveDataModel.parse_obj, input_data)
+
+    def tearDown(self) -> None:
+        BaseMario.should_raise_error = False
+        CapeMario.should_raise_error = False
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/tests/model/test_recursive_data.py:172]: The problem is that we have to set it before the first run and we want to set it back to false after every run. We could get rid of the tearDown call. However, it might be in an incorrect state afterwards. If you have a better solution to tell the class that the exception should be raised I would love to improve that.

**Comment by JulianRoesner** [comment]
# Short Description
There is more. This now also adapts the eligibility steps. They use the eligibility data models to decide what step to show next or if the data provided by the user is enough to decide if they can use the Steuerlotse and if not show the correct error page.

# Changes
- Add new eligibility steps (input, decision and failure). Input is only needed for the first eligibility step all the other input steps are decision step (deciding which step to show next)
- Add all the eligibility steps and failure pages
- Adapt the design of the success and failure page and the input forms for the eligibility flow
- Add additional title, intro and form if the user is using the Steuerlotse with their partner. These are used depending on the result of the number of users function
- Set the back link according to the input data. For some of the steps in the logic we want to link back to the correct last input step
- Add a lot of tests

# Feedback
- Do you see missing or unnecessary tests?
- Does the whole backlink logic make sense to?
- Does the way the eligibility steps work make sense to?
- What do you think about the way I used the specific stylesheets?

**Review by frederike-ramin — COMMENTED** [review]
Alriight! I think there's maybe one bigger thing we could talk about, otherwise, it looks great. 👍 

Disclaimer: in some comments, I mostly documented my though process, so I answered them myself, I just thought maybe they're indications into what isn't 100% clear.

File: webapp/app/forms/steps/eligibility_steps.py [hunk 6]
```diff
@@ -1,22 +1,107 @@
...
+
+
+class DecisionEligibilityInputFormSteuerlotseStep(EligibilityInputFormSteuerlotseStep):
+    main_next_step_name = None
+    alternative_next_step_name = None
+
+    def __init__(self, *args, **kwargs):
+        super(DecisionEligibilityInputFormSteuerlotseStep, self).__init__(*args,
+            **kwargs,
+            )
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py]: I think this should be there by now?

File: webapp/app/forms/steps/eligibility_steps.py [hunk 7]
```diff
@@ -1,22 +1,107 @@
...
+        correct because of data input from another step, raise an IncorrectEligibilityData.
+        """
+        try:
+            self.data_model.parse_obj(stored_data)
+        except ValidationError as e:
+            if any([isinstance(raw_e.exc, InvalidEligiblityError) for raw_e in e.raw_errors]):
+                return False
+            else:
+                raise IncorrectEligibilityData
+        return True
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py]: Uh, that is nice. I had the exact same thought about the double `else`.  However, as this would make it more explicit, I am onboard. TIL, too :)

File: webapp/app/forms/steps/eligibility_steps.py [hunk 8]
```diff
@@ -29,106 +114,395 @@ def _main_handle(self, stored_data):
...
+            choices=[('married', _l('form.eligibility.marital_status.married')),
+                     ('widowed', _l('form.eligibility.marital_status.widowed')),
+                     ('single', _l('form.eligibility.marital_status.single')),
+                     ('divorced', _l('form.eligibility.marital_status.divorced')),
+                     ])
 
     def _main_handle(self, stored_data):
         stored_data = super()._main_handle(stored_data)
-        self.render_info.additional_info['next_button_label'] = _('form.eligibility.send-button')
+        stored_data = dict(filter(lambda elem: elem[0] == 'marital_status_eligibility', stored_data.items()))
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py:202]: Only the MaritalStatusInputStep does not have a recursive data_model. Therefore, it is not possible to use the same function as the one you are mentioning (which isn't part of the parent step of this one anyway). However, we could restructure this a little bit and then have it here in the same way. I did not do this, because this is the first step and works a little different as it does not rely on the others. 

While thinking about it, we could put this into the parent class. and then have the decision input step override the function?

File: webapp/app/forms/steps/eligibility_steps.py [hunk 9]
```diff
@@ -29,106 +114,395 @@ def _main_handle(self, stored_data):
         return stored_data
 
 
-class EligibilityIncomesFormSteuerlotseStep(FormSteuerlotseStep):
-    name = 'incomes'
-    title = _l('form.eligibility.income-title')
-    intro = _l('form.eligibility.income-intro')
-    template = 'eligibility/form_incomes.html'
+class MaritalStatusInputFormSteuerlotseStep(EligibilityInputFormSteuerlotseStep):
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py:176]: I am not really sure, if I get your question?

You are right, if this isn't the first step anymore, we would have to change this step.

File: webapp/app/forms/steps/eligibility_steps.py [hunk 10]
```diff
@@ -29,106 +114,395 @@ def _main_handle(self, stored_data):
...
+    name = "marital_status"
+    title = _('form.eligibility.marital_status-title')
+    intro = _('form.eligibility.marital_status-intro')
+    next_steps = {
+        'married': "separated",
+        'widowed': "single_alimony",
+        'single': "single_alimony",
+        'divorced': "divorced_joint_taxes",
+    }
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py:184]: Let's talk about it again, because I think the steps changed a little. That said, we could also use a dict but this would make it a little less obvious that we are only deciding between two different steps

File: webapp/app/forms/steps/eligibility_steps.py [hunk 11]
```diff
@@ -29,106 +114,395 @@ def _main_handle(self, stored_data):
...
-            return render_template(
-                'eligibility/display_success.html',
-                title=_('form.eligibility.success-title'),
-                render_info=self.render_info,
-                header_title=_('form.eligibility.header-title')
-            )
+class MarriedJointTaxesDecisionEligibilityInputFormSteuerlotseStep(DecisionEligibilityInputFormSteuerlotseStep):
+    name = "married_joint_taxes"
+    alternative_next_step_name = MarriedJointTaxesEligibilityFailureDisplaySteuerlotseStep.name
+    title = _('form.eligibility.joint_taxes-title')
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py]: 👍

File: webapp/app/forms/steps/steuerlotse_step.py [hunk 12]
```diff
@@ -116,38 +121,43 @@ def create_form(self, request, prefilled_data):
...
-            if any([field.startswith(data_field_prefix) for data_field_prefix in data_field_prefixes]):
-                stored_data.pop(field)
-        return stored_data
+    def _delete_dependent_data(data_field_identifier: list, stored_data: dict):
+        """This method filters the stored data. It deletes all the elements where the key includes the
+        data_field_identifier. """
+        return dict(filter(
+            lambda elem: not any([elem[0].startswith(data_field_prefix) for data_field_prefix in data_field_identifier])
+                     and not any([elem[0].endswith(data_field_prefix) for data_field_prefix in data_field_identifier]),
+            stored_data.items()))
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/steuerlotse_step.py]: > I could imagine that pseron_a_religion and steuerminderung_religion could become a problem)
Only if you are looking for religion, correct?

Oh, I had another look at your proposed changes and with them you would be correct. Actually, I think I would need to change the description: `This method filters the stored data. It deletes all the elements where the key begins or ends with the data_field_identifier.` and then maybe `data_fied_subfix`?


Nevertheless, we do not need to change this function because we now do delete the data differently for the eligibility steps.

File: webapp/app/model/eligibility_data.py [hunk 13]
```diff
@@ -10,41 +10,27 @@
 class InvalidEligiblityError(ValueError):
     """Exception thrown in case the eligibility check failed."""
     _ERROR_MESSAGES = {
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/model/eligibility_data.py]: They are gone now

File: webapp/app/model/eligibility_data.py [hunk 14]
```diff
@@ -310,11 +296,11 @@ def one_previous_field_has_to_be_set(cls, v, values):
 
 class MoreThanMinimalInvestmentIncome(RecursiveDataModel):
     has_investment_income: InvestmentIncomeEligibilityData
-    minimal_investment_income: str
+    minimal_investment_income_eligibility: str
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/model/eligibility_data.py:272]: Exactly. We would not need this anymore, as we now have them stored in their own part of the session cookie.

File: webapp/app/model/recursive_data.py [hunk 15]
```diff
@@ -24,24 +42,41 @@ def _update_input(self, data: Any) -> Any:
...
+        non_found = False
         for field in fields:
             if issubclass(field.type_, BaseModel):
                 self.__class__._previous_fields.append(field.name)
-                self._set_data_for_previous_field(enriched_data, field.name, field.type_)
+                if self._set_data_for_previous_field(enriched_data, field.name, field.type_):
+                   non_found = True
 
+        if non_found:
+            print(self.__class__.__name__)
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/model/recursive_data.py]: This was only added for testing purposes. That said, we have a slight problem with the recursive data model, as it might be harder to understand why something failed in the upper chain

File: webapp/app/model/recursive_data.py [hunk 16]
```diff
@@ -24,24 +42,41 @@ def _update_input(self, data: Any) -> Any:
...
     def _set_data_for_previous_field(enriched_data, field_name, field_type):
         try:
             possible_data = field_type.parse_obj(enriched_data).dict()
             enriched_data[field_name] = possible_data
-        except ValidationError:
-            return
+            return True
+
+        except ValidationError as e:
+            return False
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/model/recursive_data.py]: You mean with:
`else:
    enriched_data[field_name] = possible_data`

File: webapp/app/model/recursive_data.py [hunk 17]
```diff
@@ -24,24 +42,41 @@ def _update_input(self, data: Any) -> Any:
...
             raise MissingError
-        return v
\ No newline at end of file
+        return v
+
+    @classmethod
+    def get_all_potential_keys(cls):
+        potential_keys = []
+        potential_keys += cls.schema().get('properties').keys()
+        definitions = cls.schema().get('definitions')
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/model/recursive_data.py:75]: Yes, when I was developing this, I started with a recursive function and then realized I do not need it 🤣

File: webapp/tests/forms/steps/test_steuerlotse_step.py [hunk 18]
```diff
@@ -174,6 +174,37 @@ def test_if_session_data_and_default_data_different_then_update_session_data(sel
...
+    def test_if_session_data_in_incorrect_identifier_then_return_only_data_from_correct_identifier(self):
+        form_data = {"brother": "Luigi"}
+        incorrect_identifier = {"enemy": "Bowser"}
+        expected_data = {**form_data}
+
+        with app.app_context() and app.test_request_context() as req:
+            req.session = SecureCookieSession({SteuerlotseStep.session_data_identifier: create_session_form_data(form_data),
+                                               "INCORRECT_IDENTIFIER": create_session_form_data(incorrect_identifier)})
+            steuerlotse_step_with_default_data = SteuerlotseStep(endpoint=self.endpoint_correct,
+                                                                 default_data={}, header_title=None,
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/tests/forms/steps/test_steuerlotse_step.py:186]: Yes, I copied the tests and I overlooked this

File: webapp/app/forms/steps/eligibility_steps.py [hunk 19]
```diff
@@ -1,22 +1,107 @@
...
+        return stored_data
 
-from app.model.eligibility_data import ExpectedEligibility, InvalidEligiblityError
+    def delete_not_dependent_data(self, stored_data):
+        return dict(filter(lambda elem: elem[0] in self.data_model.get_all_potential_keys(), stored_data.items()))
+
+    def _validate(self, stored_data):
+        """
+        Method to find out whether the data entered by the user is eligible for this step or not. If the data is not
+        correct because of data input from another step, raise an IncorrectEligibilityData.
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py:146]: `The step might depend on data from the steps before. If that data is not correct, in other words if the user could not have come from an expected step to this step by entering the correct data, raise an IncorrectEligibilityData.`

Does it now make more sense?

**Review by frederike-ramin — COMMENTED** [review]
Here goes review 3. :D

File: webapp/app/forms/steps/eligibility_steps.py [hunk 20]
```diff
@@ -22,38 +24,92 @@ class IncorrectEligibilityData(Exception):
...
+            form_multiple=self.InputMultipleForm,
             endpoint=endpoint,
             header_title=_('form.eligibility.header-title'),
             **kwargs,
-            )
+        )
+
+    def _main_handle(self, stored_data):
+        stored_data = super()._main_handle(stored_data)
+        self.set_correct_previous_link(stored_data)
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py:95]: That is funny. I had you in the back of my mind  that we do not want to override pre or post_handle but mainly the main handle for specifics.

We still can override the prev_url in the main_handle of inheriting steps. They could just override the main_handle function. However, if we would put that in the post_handle function we would now always have to override the post_handle function, too.

So, I think either _pre_handle or _main_handle make the most sense to me.

File: webapp/app/forms/steps/eligibility_steps.py [hunk 21]
```diff
@@ -22,38 +24,92 @@ class IncorrectEligibilityData(Exception):
...
+    def set_correct_previous_link(self, stored_data):
+        if self.previous_steps:
+            back_link_url = None
+
+            if len(self.previous_steps) == 1:
+                back_link_url = self.url_for_step(self.previous_steps[0].name)
+            else:
+                for previous_step in self.previous_steps:
+
+                    if not issubclass(previous_step, DecisionEligibilityInputFormSteuerlotseStep):
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py:107]: Basically I am testing for the maritalinputSteuerlotseStep as this one would not have a data model

File: webapp/app/forms/steps/eligibility_steps.py [hunk 22]
```diff
@@ -22,38 +24,92 @@ class IncorrectEligibilityData(Exception):
...
+                        back_link_url = self.url_for_step(previous_step.name)
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py:108]: If the user goes through the steps in the intended way, there should always only be one previous step. So, we can either break here or we do not break and take the last input, which was intended by me. There is no specific reasoning behind that. I set the steps always in order. So, if for some reason the data would match more than one step (e.g. user setting their own data), we would get the last possible step in that way.

TLDR: If you would like to have a break, you could have one 😄

File: webapp/app/forms/steps/eligibility_steps.py [hunk 23]
```diff
@@ -152,357 +215,504 @@ class SeparatedEligibilityInputFormSteuerlotseStep(DecisionEligibilityInputFormS
...
         kwargs['prev_step'] = ForeignCountriesDecisionEligibilityInputFormSteuerlotseStep
-        super(EligibilitySuccessDisplaySteuerlotseStep, self).__init__(endpoint=endpoint, header_title=_('form.eligibility.header-title'), **kwargs)
+        super(EligibilitySuccessDisplaySteuerlotseStep, self).__init__(endpoint=endpoint,
+                                                                       header_title=_('form.eligibility.header-title'),
+                                                                       **kwargs)
+
+    def _main_handle(self, stored_data):
+        stored_data = super()._main_handle(stored_data)
+
+        dependent_notes = []
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py:708]: Because they depend on the user input?

File: webapp/app/templates/components.html [hunk 24]
```diff
@@ -70,7 +70,7 @@
     {%- set detail = field.render_kw['detail'] -%}
 
     {% if field.type == 'RadioField' %}
-        {% if label.text %} {# Only show legend if there is a legend text to show #}
+        {% if label.text and not hide_label and not (field.render_kw and 'hide_label' in field.render_kw and field.render_kw["hide_label"]) %} {# Only show legend if there is a legend text to show #}
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/templates/components.html:73]: We set them to true but they do not hide the label as the label text is set here. I have to check again.

File: webapp/tests/forms/steps/test_eligibility_steps.py [hunk 25]
```diff
@@ -72,10 +76,138 @@ def test_if_start_step_then_return_redirect_step(self):
...
+        input_data = {'marital_status_eligibility': 'married',
+                      'separated_since_last_year_eligibility': 'no',
+                      'joint_taxes_eligibility': 'yes', }
+        num_of_users = EligibilityStepSpecificsMixin().number_of_users(input_data)
+
+        self.assertEqual(2, num_of_users)
+
+    def test_if_data_incorrect_then_return_1(self):
+        input_data = {'marital_status_eligibility': 'widowed'}
+        num_of_users = EligibilityStepSpecificsMixin().number_of_users(input_data)
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/tests/forms/steps/test_eligibility_steps.py:99]: I do not think that we exclude them specifically. We just do not ask them if they want to use the Zusammenveranlagung.

File: webapp/tests/forms/steps/test_eligibility_steps.py [hunk 26]
```diff
@@ -72,10 +76,138 @@ def test_if_start_step_then_return_redirect_step(self):
...
+            step = EligibilityInputFormSteuerlotseStep(endpoint="lotse")
+            previous_url = step.url_for_step(self.NormalStep.name)
+            step.name = "CURRENT"
+            step._pre_handle()
+            step.render_info.prev_url = previous_url
+            step.previous_steps = [self.InValidPreviousStep,  self.ValidPreviousStep, self.NormalStep]
+
+            step.set_correct_previous_link(stored_data)
+
+            self.assertEqual(previous_url, step.render_info.prev_url)
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/tests/forms/steps/test_eligibility_steps.py:203]: s. comment above

File: webapp/app/forms/steps/eligibility_steps.py [hunk 27]
```diff
@@ -62,16 +118,19 @@ class DecisionEligibilityInputFormSteuerlotseStep(EligibilityInputFormSteuerlots
 
     def __init__(self, *args, **kwargs):
         super(DecisionEligibilityInputFormSteuerlotseStep, self).__init__(*args,
-            **kwargs,
-            )
+                                                                          **kwargs,
+                                                                          )
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py]: I think I do not really get your comment. Do you just want one line?

BTW: It would help me a lot if you could (especially for something like this) include a few more lines around the important part for context 😬

File: webapp/app/forms/steps/eligibility_steps.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/forms/steps/eligibility_steps.py:146]: Noice! Yes"

File: webapp/app/forms/steps/eligibility_steps.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/forms/steps/eligibility_steps.py]: I think I was really just saying that I don't like **kwargs on the next line. ;)
(and also, I don't think this has changed)

File: webapp/app/model/recursive_data.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/model/recursive_data.py]: I mean 
```
        try:
            possible_data = field_type.parse_obj(enriched_data).dict()
        except ValidationError as e:
            return False
        else:
            enriched_data[field_name] = possible_data
            return True
```

File: webapp/app/forms/steps/eligibility_steps.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/forms/steps/eligibility_steps.py:95]: Oh, of course, I got it the complete wrong way around. 🙈 
And I think we can override pre_ and post_handle, just not for the last inheritance step, if that makes sense?
But with your explanation I actually think it's fine to leave it here^^

File: webapp/app/forms/steps/eligibility_steps.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/forms/steps/eligibility_steps.py]: Yes, just one line, exactly.

It's weird, but I feel like GitHub did that for me for a while there, they just seem to have stopped. But maybe I'm wrong. I'll try my best. :)

File: webapp/app/forms/steps/eligibility_steps.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/forms/steps/eligibility_steps.py:708]: pfff..sure. :D Not really convinced, but fine for me^^

File: webapp/app/forms/steps/steuerlotse_step.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/forms/steps/steuerlotse_step.py]: Add prefix + postfix parameters

File: webapp/app/forms/steps/eligibility_steps.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/forms/steps/eligibility_steps.py:108]: Yes I'd like to have a KitKat 🍪

File: webapp/app/templates/components.html (hunk 24 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/templates/components.html:73]: We talked about it and will set label='' on the fields.

File: webapp/app/forms/steps/eligibility_steps.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py]: This can also go now that maritalinputSteuerlotseStep is a multiple decision step

File: webapp/app/forms/steps/eligibility_steps.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py]: s. other answer about marital status input step

File: webapp/app/forms/steps/eligibility_steps.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py]: s. other answer about marital status input step

File: webapp/app/forms/steps/eligibility_steps.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py]: We've decided to add an additional step for multiple decisions and use that one for the marital status input. This step will share a lot more with the decision steps then.

File: webapp/app/model/recursive_data.py [hunk 28]
```diff
@@ -9,7 +9,22 @@ class PreviousFieldsMissingError(MissingError):
     """Raised in case all the previous fields are missing."""
 
 
-class RecursiveDataModel(BaseModel):
+class PotentialDataModelKeysMixin:
```
Comment (frederike-ramin) [inline webapp/app/model/recursive_data.py:12]: Maaayybe 
```suggestion
class PotentialDataKeysModelMixin:
```

File: webapp/app/templates/components.html [hunk 29]
```diff
@@ -35,52 +35,64 @@
...
+                {{ field_label(subfield, class="col-sm-10 col-form-label", position_details_after=position_details_after) }}
             </div>
         {% endfor %}
+        {%- if position_details_after %}
+            {%- set detail = field.render_kw['detail'] -%}
+            {%- if detail %}
+                {{ details(detail.title, [detail.text], details_id=field.id) }}
+            {% endif %}
+        {% endif %}
         {{ field_errors(field) }}
```
Comment (frederike-ramin) [inline webapp/app/templates/components.html:69]: Should we maybe make this a separate macro? like _detail() or something? then we'd only need to change one place.

Comment (frederike-ramin) [reply webapp/app/templates/components.html:69]: And is it a decision to add it do the separated field but not to consent or checkbox fields?

File: webapp/app/templates/components.html [hunk 30]
```diff
@@ -35,52 +35,64 @@
...
 
-{% macro field_label(field, class="") %}
+{% macro field_label(field, class="", position_details_after=False) %}
     {%- set label = field.label -%}
     {%- set help = field.render_kw['help'] -%}
     {%- set detail = field.render_kw['detail'] -%}
 
-    {% if field.type == 'RadioField' %}
-        {% if label.text and not hide_label and not (field.render_kw and 'hide_label' in field.render_kw and field.render_kw["hide_label"]) %} {# Only show legend if there is a legend text to show #}
+    {% if field.type == 'RadioField' or field.type == 'UnlockCodeField' %}
```
Comment (frederike-ramin) [inline webapp/app/templates/components.html]: Not sure this is on purpose? If it's not on main, it probably shouldn't be here?

**Review by frederike-ramin — COMMENTED** [review]
Nice code!

I tried it out and found:
- the success_icon is I think not pushed? At least, I only get the alt-text (also on the failure page - is that on purpose?
- also, on the failure page I see the heading twice: "Sie können den Steuerlotsen zurzeit nicht für Ihre Steuererklärung nutzen." - not sure this is on purpose

Otherwise it looks great^^

File: webapp/app/forms/steps/steuerlotse_step.py [hunk 31]
```diff
@@ -150,14 +150,17 @@ def render(self, **kwargs):
...
-    def _delete_dependent_data(data_field_identifier: list, stored_data: dict):
+    def _delete_dependent_data(stored_data: dict, pre_fixes:list=None, post_fixes:list=None):
         """This method filters the stored data. It deletes all the elements where the key includes the
         data_field_identifier. """
-        return dict(filter(
-            lambda elem: not any([elem[0].startswith(data_field_prefix) for data_field_prefix in data_field_identifier])
-                         and not any(
-                [elem[0].endswith(data_field_prefix) for data_field_prefix in data_field_identifier]),
-            stored_data.items()))
+        filtered_data = stored_data
```
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/steuerlotse_step.py:156]: I can do the copy here; I did not do so because of performance . It is basically to ensure that we return something if neither prefix nor postfix are set.

File: webapp/app/model/recursive_data.py (hunk 28 above)
Comment (JulianRoesner) [reply webapp/app/model/recursive_data.py:12]: I did it the other way around, because these are the keys of the data model

File: webapp/app/templates/components.html (hunk 30 above)
Comment (JulianRoesner) [reply webapp/app/templates/components.html]: What is this. You mean UnlockCodeField?

Comment (JulianRoesner) [reply webapp/app/templates/components.html]: This was from one of the merges

File: webapp/app/forms/steps/steuerlotse_step.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (frederike-ramin) [reply webapp/app/forms/steps/steuerlotse_step.py:156]: Yes, makes sense. just thinking that we could also just keep it as stored_data and modify that without possibly confusing with the re-assignment. But completely open to you, no strong opinion.

File: webapp/app/model/recursive_data.py (hunk 28 above)
Comment (frederike-ramin) [reply webapp/app/model/recursive_data.py:12]: Ahh..sure!

File: webapp/app/templates/components.html (hunk 30 above)
Comment (frederike-ramin) [reply webapp/app/templates/components.html]: Yes. I mean exactly that. Just thought, if it's not on main and you didn't change it on purpose, it's probably an error.

File: webapp/app/forms/steps/steuerlotse_step.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (JulianRoesner) [reply webapp/app/forms/steps/steuerlotse_step.py:157]: Yes, but I would then feel not so good about it, because we do not want to modify the input but only the returned result here.

File: webapp/app/templates/components.html (hunk 30 above)
Comment (JulianRoesner) [reply webapp/app/templates/components.html]: Yes, if I am not totally wrong, it is on main and I do not know why GitHub is showing this change here.

File: webapp/app/templates/components.html (hunk 29 above)
Comment (JulianRoesner) [reply webapp/app/templates/components.html:73]: > Should we maybe make this a separate macro? like _detail() or something? then we'd only need to change one place.

But how would that work? If we do not set it at this point we do not set it?

They do not have a details component and therefore I did not add one.

Comment (JulianRoesner) [reply webapp/app/templates/components.html]: Okay, I've added a macro for that
